Concretization Cache: Refactor + manifest testing

johnwparent opened the pull request:
#48198 Introduces the concretization cache, and associated manifest. However it only added testing for the cache itself, not the manifest or cleanup procedures.

This PR:

- [x] Adds tests for the manifest and cleanup
- [x] Refactors the manifest to update correctly
- [x] Refactors the cache to store base32 hashes like the rest of Spack's hashes
- [x] gzips the cache entries for a smaller memory footprint

`FileCache` class was refactored into an abstract base `Cache` class which provides the general cache read/write transaction interface, and a `FileCache` and `DirectoryCache` child classes, `FileCache` which should be functionally identical to the `FileCache` class pre refactor, deals with caches at the individual file level, and `DirectoryCache`, which is mean to provide a directory level or "bucket" caching level.

johnwparent committed:
General refactors and improvements to the concretization cache

johnwparent committed:
Add test for cache cleanup

johnwparent committed:
Style

johnwparent committed:
refactor tests

johnwparent commented:
@spackbot fix style

johnwparent committed:
[@spackbot] updating style on behalf of johnwparent

johnwparent committed:
Clear first line of new manifest

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -824,28 +834,56 @@ def _lock_prefix_from_cache_path(self, cache_path: str):
-            f.write(new_stats)
+                new.write(f"{entry_path.name} {entry_bytes}\n")
+            new.seek(0, io.SEEK_SET)
+            new_stats = f"{count} {cache_bytes}\n"
```
This in-place mutation does not work because line length is variable. If it increases, you overwrite the newline character, and the file gets corrupted. I think you should either work with a binary file (`struct.pack/struct.unpack*`), or rewrite the entire file (probably what you wanted to avoid).

Note that if the idea is to do a constant time write operation (update header & append single entry), then `FileCache` is the wrong choice, since it copies the file to a temporary which you then mutate in-place. If you want to do things in-place, you could work with the lock file directly.

haampie replied:
Also what's missing is some version field in the header to describe the metadata file format, which is always helpful in case we make a forward incompatible change.

johnwparent replied:
> what's missing is some version field in the header to describe the metadata file format

Good idea, adding now. 

> This in-place mutation does not work because line length is variable.

🤦 

> Note that if the idea is to do a constant time write operation (update header & append single entry), then FileCache is the wrong choice, since it copies the file to a temporary which you then mutate in-place.

Good point, I didn't consider this when I refactored to using the FileCache, I may as well just avoid the in place write. That said, this outcome of this response conditions on our overall decision on whether or not to actually use the `FileCache`. 

Either way, I agree, it should all be binary.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -824,28 +834,56 @@ def _lock_prefix_from_cache_path(self, cache_path: str):
         manifest_file = self.root / self._cache_manifest
         manifest_file.touch(exist_ok=True)
-        with open(manifest_file, "r+", encoding="utf-8") as f:
+        with self._fc.write_transaction(self._cache_manifest) as (old, new):
```
`write_transaction` opens a file in text mode, which makes `seek` calls questionable. It does not accept byte nor character offsets, but rather whatever tell() returns, which is documented as an "opaque number" (i.e. you shouldn't pass numbers yourself)

So, if you continue to use `FileCache`, you should probably extend the API to accept a file mode and read/write in binary. But given the other review comment, maybe FileCache isn't the best choice here given that it does not allow constant time updates to the file.

johnwparent replied:
> So, if you continue to use FileCache, you should probably extend the API to accept a file mode and read/write in binary. But given the other review comment, maybe FileCache isn't the best choice here given that it does not allow constant time updates to the file.

I think, at least you and I, we are in the binary only camp, the question just remains whether or not to use `FileCache`.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4625,11 +4665,12 @@ def solve_with_stats(
+        result = self.driver.solve(
             setup, specs, reuse=reusable_specs, output=output, allow_deprecated=allow_deprecated
         )
+        CONC_CACHE.flush_manifest()
```
Shouldn't `flush` be done on every write? Then you don't need to manually sync and deal with exceptions that are raised between the writing the concretizer output to cache and updating the stats.

johnwparent replied:
Syncing slack thread here for context for other reviewers:

This was done as an effort to future proof the concretization cache against a large number of concurrent writes causing serialized cache result returns in the case where a large number of concurrent writes occur when we add the capability to have a remote concretization cache. The concern is present locally to a smaller degree when running multiple spack processes concretizing large environments with `unify: false`, particularly with  many small packages, but the primary concern is for remote caches. Particularly for smaller specs, having to wait long periods for write access to the manifest could limit the conc speedup the cache provides.

johnwparent replied:
Follow up comments from @haampie suggest the use of a `ccache` style model where we lock a "bucket" (or cache entry subdir) and track the stats at that scope, significantly limiting the occurrence of concurrent writes causing a bottleneck. The same overarching cache statistics we currently track could be computed from each bucket via a method as part of the conc cache API. 

I think this is the best path forward, currently implementing and testing.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -824,28 +834,56 @@ def _lock_prefix_from_cache_path(self, cache_path: str):
 
+    def _zip_cache_entry(self, cache_data: str):
+        """Compress cache entries via gzip"""
+        return gzip.compress(cache_data.encode("utf-8"))
```
Probably a good idea to specify the `compresslevel=`; the default is `gzip._COMPRESS_LEVEL_BEST == 9` which is very slow for little added value, level 6 is better, see the docstrings of `spack.util.archive.gzip_compressed_tarfile`

johnwparent replied:
Sounds good, thanks!

johnwparent replied:
Going off my inbox I see this comment

> You should use mtime=0, otherwise you get a different gzip header on every call and 0 cache hits.

I don't think that has an implication for cache hits since we go off the filename, but it is still best practice?

haampie replied:
Initially I did not see you checksummed the contents.

Note that if you're planning to distribute these files, you might need to do sha256sum(gzip(contents)) instead of sha256sum(contents), otherwise checksum validation requires decompression, which we try to avoid everywhere else.

johnwparent replied:
Ah ok, that makes sense, thanks for the follow up.

I like `sha256sum(gzip(contents))`, its a good idea, will refactor, was previously unaware of the convention (but also it's common sense)

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -824,28 +834,56 @@ def _lock_prefix_from_cache_path(self, cache_path: str):
+        """Compress cache entries via gzip"""
+        return gzip.compress(cache_data.encode("utf-8"))
+
+    def _unzip_cache_entry(self, cache_data: bytes):
```
it's more common to wrap the filestream (e.g. gzip.open or GzipFile(...)) so decompress happens on read(...)

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -889,7 +927,8 @@ def store(self, problem: str, result: Result, statistics: List, test: bool = Fal
             cache_dict = {"results": result.to_dict(test=test), "statistics": statistics}
-            bytes_written = new.write(json.dumps(cache_dict))
+            new_bytes = open(new.name, "wb")
+            bytes_written = new_bytes.write(self._zip_cache_entry(json.dumps(cache_dict)))
```
I think it's incorrect to open the same file a second time.

Probably easier to avoid the entire FileCache and instead create a temporary file in the dir safely (e.g. tempfile.mkstemp) followed by an atomic rename. I think we have utilities for that.

tgamblin replied:
we have `llnl.util.filesystem.write_tmp_and_move()` and a few others

kwryankrattiger replied:
The idea here is to use the `FileCache` so I can drop in a Storage object with a similar API without having to re-write too much.

johnwparent replied:
> I think it's incorrect to open the same file a second time.

I think its definitely a smell to do so, no argument there, but as we're writing binary encoded data to the entries, we need a binary writer. I for one, if we're locked into the `FileCache` approach, would like a binary file cache option.
Another alternative is adding a new API, similar to the `FileCache` but perhaps better suited to this type of operation, that way we avoid some of the cludgey-ness of this implementation approach and Ryan can still drop something in easily.

haampie replied:
Shouldn't be hard to add read/write_transaction(..., mode).

I don't really see the advantage of FileCache since it always copies the file that you're planning to update in-place. What's the reason you want to use FileCache?

johnwparent replied:
I think I touched on this in our slack conversation, and Ryan stated it above, but pretty much the only reason to use it is so that we can have a generic API that allows us to refer to either a local (managed by FileCache) or remote (managed by some remote file cache class using a compatible) cache location. It's also why the config entry is confusingly (for now) url, which is something I believed you made a point about in the other PR. 

> read/write_transaction(..., mode)

Yes, it's very very easy. ~~That said, constant time file updates with atomic writes would be nice, which isn't currently possible in the FileCache. Maybe a refactor is in order.~~

Nevermind, I'm abandoning the manifest altogether in favor of a cleanup-time computation of the cache contents, that will make things easier for Ryan and remove a good deal of the issues with this cache.

haampie replied:
hm okay, filecache is mostly for locking, that's not very relevant for remotes

kwryankrattiger replied:
I am working on a follow up that is a drop in replacement for the `FileCache` API. It is meant primarily to make adding remote caching easier later.

The web_util APIs are rather more limited and relies on global variables and is heavily tied to mirrors. Rather than a more significant refactor of that, I am working on a more generic interface for dealing with storage devices.

kwryankrattiger replied:
`FileCache` is just a placeholder for now. Writing something else the is more configurable/better suited for this is probably not a bad idea.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -824,28 +834,56 @@ def _lock_prefix_from_cache_path(self, cache_path: str):
+
+    def _unzip_cache_entry(self, cache_data: bytes):
+        """Decompresses cache entries, which are gzipped by default"""
+        return gzip.decompress(cache_data).decode("utf-8")
```
This errors for existing uncompressed cache entries.

johnwparent replied:
Yes, I need to write a handler/migration for that.

alalazo commented:
While we are fixing this, can we also add a default for:
```
            "concretization_cache": {
                "type": "object",
                "properties": {
                    "enable": {"type": "boolean"},
                    "url": {"type": "string"},
                    "entry_limit": {"type": "integer", "minimum": 0},
                    "size_limit": {"type": "integer", "minimum": 0},
                },
            },
```
in `etc/defaults`? Currently we have none, and so it's unclear which value is used by doing:
```console
$ spack config blame config
```
Also, was it discussed whether this belongs to `config` or `concretizer`?

kwryankrattiger commented:
> Also, was it discussed whether this belongs to config or concretizer?

It was, but there had since been disagreement on that. I think config makes sense as that is where we configure the other misc caches, but I don't have an extremely strong opinion being that.

tgamblin commented:
> Also, was it discussed whether this belongs to config or concretizer?

I think it would be nice if it were in `concretizer`, but you could refer to the `misc_cache` as part of the default URL.

johnwparent committed:
Better handling of intermediate binary file approach

johnwparent commented:
@spackbot fix style

johnwparent committed:
Dont read to unreadable fd

johnwparent committed:
use correct cache limit config name

wip

johnwparent committed:
cache overhaul

johnwparent committed:
Final refactor

johnwparent committed:
refactor tests post rebase

johnwparent committed:
rm unneeded test

johnwparent committed:
rm lambda cruft

johnwparent committed:
cleanup

johnwparent committed:
compat with python 3.7

johnwparent committed:
Cleanup tests + confgi

johnwparent committed:
account for differences in conc size

johnwparent committed:
rm bad docstring

johnwparent commented:
@haampie @alalazo @tgamblin @kwryankrattiger PTAL at your convenience

johnwparent commented:
@haampie @alalazo another ping

johnwparent committed:
Test conc cache roundtrip: remove forced determinism

We have actual stable asp output now, no need for the forced determinism hack

kwryankrattiger commented on etc/spack/defaults/concretizer.yaml:
```diff
@@ -89,3 +89,7 @@ concretizer:
+
+  concretization_cache:
+    enable: true
+    url: $misc_cache/concretization
```
`$misc_cache` is not a thing, the cache is getting dumped to `etc/spack/defaults/\$misc_cache/concretization/...`

kwryankrattiger replied:
Needs adding to `lib/spack/spack/util/path.py:replacements()` and in the doc-string for `substitute_config_variables`.

kwryankrattiger replied:
Or use `$user_cache_path` forgot about that one!

johnwparent replied:
Fixed

johnwparent committed:
misc_cache -> user_cache_path

kwryankrattiger reviewed (dismissed):
LGTM, tested locally and got cache hits as expected, set cache count to 4 and it correctly limits the size based on date order not list order.

@haampie @tgamblin Merge when ready.

johnwparent commented:
> LGTM, tested locally and got cache hits as expected, set cache count to 4 and it correctly limits the size based on date order not list order.
> 
> @haampie @tgamblin Merge when ready.

I know we have a test for it (and I've verified locally), but did you test the byte size limit as well?

kwryankrattiger commented:
> I know we have a test for it (and I've verified locally), but did you test the byte size limit as well?

I just ran the test and byte limits don't seem to be working correctly. Setting to 200000 and I have an old concretization cached that is >400k.

johnwparent committed:
Remove dropped race condition handling

kwryankrattiger commented:
Okay. User error! Size pruning works :) 

Some other things I noted:

* I do get stale locks that are not cleaned up for the buckets after they are removed. That should probably be cleaned up when buckets are deleted.
* if the resulting cache item is > the the `size_limit` it is immediately purged, AND it also purges older cache entries that could have persisted under the size limit. I don't think we need handle that as a special case, I just noticed it in testing and wasn't expecting it.

johnwparent commented:
> Okay. User error! Size pruning works :)

Ok great.


> * I do get stale locks that are not cleaned up for the buckets after they are removed. That should probably be cleaned up when buckets are deleted.

There's currently a race between cleanups and init entries that allows for a lockfile to be created but not yet associated with any writes, and a cleanup to purge the bucket before a writing process can take a lock. Working out best way to resolve that now.

> * if the resulting cache item is > the the `size_limit` it is immediately purged, AND it also purges older cache entries that could have persisted under the size limit. I don't think we need handle that as a special case, I just noticed it in testing and wasn't expecting it.

I think the pruning logic could use an overhaul, but perhaps as a followup.

johnwparent committed:
Cleanup lockfiles

johnwparent committed:
concretization cache: Improve handling of lockfile cleanup

johnwparent committed:
Concretization cache tests: add lockfile cleanup unit test

johnwparent committed:
style

johnwparent commented:
@kwryankrattiger @tgamblin @haampie PTAL at your convenience

johnwparent committed:
A legitimate skip test on Windows

johnwparent committed:
style

johnwparent committed:
Concretization Cache: add test for migration from uncompressed to compressed entries

Previous Versions of the concretization cache were uncompressed. We now compress them to save space.
Explicitly test that the new concretization cache can handle both compressed and uncompressed entries.

johnwparent committed:
typo

johnwparent committed:
Add file cache tests for directory caceh

johnwparent committed:
refactor directory cache test

johnwparent committed:
style

johnwparent committed:
Concretizer cache test: add encoding

johnwparent committed:
rm bad path
